File: lib/domain/cultivation_plan/errors/entry_schedule_optimization_error.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Errors
      # EntryScheduleOptimizeInteractor が optimize_period 失敗を Result に写すためのドメイン例外
      class EntryScheduleOptimizationError < StandardError
        attr_reader :error_key

        def initialize(error_key, message = nil)
          @error_key = error_key
          super(message || error_key.to_s)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/errors/entry_schedule_optimization_error.rs
//! Ruby: `Domain::CultivationPlan::Errors::EntryScheduleOptimizationError`

use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub struct EntryScheduleOptimizationError {
    pub error_key: String,
    pub message: String,
}

impl EntryScheduleOptimizationError {
    pub fn new(error_key: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            error_key: error_key.into(),
            message: message.into(),
        }
    }
}

impl fmt::Display for EntryScheduleOptimizationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for EntryScheduleOptimizationError {}
